docs/design/evm-versioning.md

```diff
@@ -0,0 +1,62 @@
+security demands it.
+
+In order for off-chain evaluation of these prior EVMs to work we also need to preserve a way to re-create the EVM
+classes in such a way so that the other instances can re-create the execution at
+
+## Goals
+
+- Allow EVM "versions" to be switched by dynamic system properties
+- Allow EVM infrastructure to be committed to production code without requiring it to be activated at consensus.
+- Allow older versions of the EVM to remain accessible for
+
+## Non Goals
+
+- Switching the EVM version dynamically at runtime. EVM will be locked in at system startup
+
+## Architecture
+
+First, the EVM object will be created by dagger injection. Previously these were constructed in the EVM Processor for
+each HederaTransaction Type, where Dagger provided the operations and gas calculator and the EVM was created in situ.
+
+Second, the dagger injection will be done via a map of providers, keyed by a text string corresponding to the advertised
+version of the EVM. Using a provider will result in just-in-time construction of the object and will not cause multiple
+versions of the EVM to exist in-memory.  (Done via `@IntoMap` bindings for the EVM and injected
+as `Map&lt;String, Provider&lt;EVM>>`).
+
+Finally, the required differences will be populated into separate submodules, using a version appropriate dagger
+qualifier.
+
+## Non-Functional Requirements
+
+A standard dynamic property will be used to configure the EVM at startup (`contracts.evm.version`) and will be defaulted
+to the last activated version on mainnet in the event the value is not set.
+
+EVM versions will follow the format `v&lt;major>.&lt;minor>`, corresponding to the released version of hedera. for
+example, the earliest version supported by this regime is `v0.30` with a planned update for `v0.31`. Not every hedera
+version will have a new EVM version. Only when EVM compatability is impacted for object replay will a new version be
+set.
+
+It is expected that there will be a new version for each major Ethereum Mainnet hard fork. A table will be kept here to
+document which major hardfork corresponds to each internal version.
+
+|  Hedera Version | Ethereum Fork | Comments                                                                              |
+|----------------:|:--------------|:--------------------------------------------------------------------------------------|
+|         `v0.30` | `London`      |                                                                                       |
+|         `v0.31` | `Paris`       | Replaces `DIFFICULTY` with `RANDAO`, removes errors from Invalid Solidity Addresses   |
+|               ? | `Shanghai`    | Expected Q2 2023 or later                                                             |
+
+## Open Questions
+
+The exact timing of versions that correspond to Ethereum Mainnet forks is out of scope of this document
+
+## Acceptance Tests
```

Comment:
Should call out a version verification scenarios that should be captured.

Reply:
updated.